prusti-specs/src/extern_spec_rewriter.rs

```diff
@@ -201,76 +136,470 @@ fn rewrite_method_inputs(item_ty: &syn::Type, method: &mut ImplItemMethod) ->
-/// but then need to pass those as arguments: `SomeStruct<...>`. This function translates from
-/// the syntax of one to the other; e.g. `<T: Bound, 'l: Bound, const C: usize>` -> `<T, 'l, C>`
-pub fn rewrite_generics(gens: &syn::Generics) -> syn::AngleBracketedGenericArguments {
-    let args: Vec<syn::GenericArgument> = gens.params.clone().iter().map(|gp| {
-        let ts = match gp {
-            syn::GenericParam::Type(syn::TypeParam { ident, .. }) |
-            syn::GenericParam::Const(syn::ConstParam { ident, .. }) => ident.into_token_stream(),
-            syn::GenericParam::Lifetime(ld) => ld.lifetime.clone().into_token_stream(),
+/// Rewriting of extern specs on traits
+pub mod traits {
+    use super::*;
+
+    type AssocTypesToGenericsMap<'a> = HashMap<&'a syn::Ident, syn::TypeParam>;
+
+    /// Generates a struct for a `syn::ItemTrait` which is used for checking
+    /// compilation of external specs on traits.
+    ///
+    /// Given an extern spec for traits
+    /// ```rust
+    /// #[extern_spec]
+    /// trait SomeTrait where Self: OtherTrait {
+    ///     type ArgTy;
+    ///     type RetTy;
+    ///
+    ///     fn foo(&self, arg: Self::ArgTy) -> Self::RetTy;
+    /// }
+    /// ```
+    /// it produces a s struct
+    /// ```rust
+    /// struct Aux<TSelf, TArgTy, TRetTy> {
+    ///     // phantom data for TSelf, TArgTy, TRetTy
+    /// }
+    /// where TSelf: SomeTrait<ArgTy = TArgTy, RetTy = TRetTy> + OtherTrait
+    /// ```
+    /// and a corresponding impl block with methods of `SomeTrait`.
+    ///
+    pub fn handle_extern_spec_trait(item_trait: &syn::ItemTrait) -> syn::Result<TokenStream> {
+        validate_macro_usage(item_trait)?;
+
+        let generated_struct = generate_new_struct(item_trait)?;
+
+        let trait_impl = generated_struct.generate_impl();
+        let new_struct = generated_struct.generated_struct;
+        Ok(quote_spanned! {item_trait.span()=>
+                #new_struct
+                #trait_impl
+        })
+    }
+
+    /// Returns an error when the macro was used in a wrong way on the trait
+    fn validate_macro_usage(item_trait: &syn::ItemTrait) -> syn::Result<()> {
+        // Default methods not allowed
+        for item in item_trait.items.iter() {
+            if let syn::TraitItem::Method(item_method) = item {
+                if item_method.default.is_some() {
+                    return Err(syn::Error::new(
+                        item_method.default.as_ref().unwrap().span(),
+                        "Default methods in external trait specs are disallowed \
+                        because external specs are always trusted",
+                    ));
+                }
+            }
+        }
+
+        // Generics not supported
+        if item_trait.generics.params.len() > 0 {
+            return Err(syn::Error::new(
+                item_trait.span(),
+                "Generics in external trait specs are not supported",
```

Comment:
Why? Is it not just a matter of parsing the list on the declaration, stripping out the bounds(?), and prepending it `where TSelf: SomeTrait<(here), AssocT1 = ...>`?